pcdsdevices/state.py

```diff
@@ -0,0 +1,376 @@
+        if not(issubclass(enum, Enum)):
+            err = ('Expected a subclass of enum.Enum, but recieved type {}')
+            raise TypeError(err.format(type(self._states_enum)))
+
+        super().__init__(read_pv, write_pv=read_pv+":GO", string=True,
+                         name=name, parent=parent, **kwargs)
+
+        self.enum = enum
+        self._enum_strs = [self.unk] + [state.name for state in enum]
+        self.enum_vals = [0] + [state.value for state in enum]
+        self.good_vals = self._enum_strs[1:] + self.enum_vals[1:]
+
+    @property
+    def enum_strs(self):
+        return self._enum_strs
+
+    def check_value(self, value):
+        if not isinstance(value, (str, int)):
+            raise TypeError('Valid states must be of type str or int')
+        elif value in ('Unknown', 0):
+            raise ValueError('Cannot move to the Unknown state (0)')
+        elif value not in self.good_vals:
+            err = ('{0} is not a valid state for {1}. Valid state names are: '
+                   '{2}, and their corresponding values are {3}.')
+            raise ValueError(err.format(value, self.name,
+                                        self._enum_strs, self.enum_vals))
+
+
+class StatePositioner(Device, PositionerBase):
+    """
+    Base class for positioners with a discrete set of named positions
+    facilitated by the lcls states record.
+    """
+    state = Component(StateSignal, '', enum='_states_enum')
+    readback = FormattedComponent(EpicsSignalRO, '{self._readback}')
+
+    _states_enum = None  # Override with an Enum to define your states
+    _states_alias = {}   # Override with {'ALIAS': 'STATE'}
+
+    SUB_STATE = 'state'
+    _default_sub = SUB_STATE
+
+    _default_read_attrs = ['state', 'readback']
+    _default_configuration_attrs = []
+
+    def __init__(self, prefix, *, name, **kwargs):
+        some_state = next(a for a in self._states_enum).name
+        self._readback = '{}:{}_CALC.A'.format(prefix, some_state)
+        super().__init__(name=name, **kwargs)
+
+        self.prefix = prefix
+
+        self._inverse_alias = {value: key for key, value in
+                               self._states_alias.items()}
+        self._has_subscribed_state = False
+        self._has_subscribed_readback = False
+
+    def move(self, position, moved_cb=None, timeout=None, wait=False):
+        """
+        Move to the desired state.
+
+        Parameters
+        ----------
+        position: int or str
+            The enumerate state or the corresponding integer
+
+        moved_cb: function, optional
+            moved_cb(obj=self) will be called at the end of motion
+
+        timeout: int or float, optional
+            Move timeout in seconds
+
+        wait: bool, optional
+            If True, do not return until the motion has completed.
+        """
+        status = self.set(position, moved_cb=moved_cb, timeout=timeout)
+        if wait:
+            status_wait(status)
```

Comment:
This should return the status?

Reply:
This was a concious choice. I'd rather create a convention where `move` and other convenience methods are designed for the command-line user, where having the status pop up is annoying as per scientist feedback. Then we reserve the `set` method for the raw "just do the thing and give me a status object". This creates a clear distinction as to what belongs in each method and creates a path to do status bars in move, etc.

I realize as typing this that I took this too far in `valve.py` and probably messed up part of the `lightpath` interface. This should be revisited before the merge.

Reply:
Yeah. We need those `insert`, `remove` methods to return a Status object. I'm fine with this convention.

Reply:
This morning I realized that this crusade was kind of overkill. We should return `Status` objects from `move` because that's what `ophyd` does, but then reserve the `mv`, `mvr`, etc. methods for the bells and whistles and provide them with multi-inheritance from a simple interface class that references `set` or `move`. I still think `set` shouldn't have any special options like `wait` which belong in `move`.